### PR TITLE
[incubator/keycloak] Add high availability and various improvements

### PR DESCRIPTION
**Pull request by unguiculus**
* Add various improvements, such as node selector, tolerations, and pod anti-affinity
* Switch to StatefulSet due to length restrictions for WildFly node identifier and
  for potential advantages for cluster discovery when names pod stable
* Add high availability and clustering using JDBC_PING discovery for JGroups
* Add customizability via WildFly CLI
* Improve ingress configuration
* Add test

### CODE REVIEW

**Comment by unguiculus** [comment]
cc @nexeck

File: incubator/keycloak/templates/NOTES.txt [hunk 1]
```diff
@@ -5,40 +5,47 @@
...
   * From outside the cluster, the URL is:
-{{- if .Values.keycloak.ingress.tls }}
+
+  {{- if .Values.keycloak.ingress.tls }}
     https://{{ .Values.keycloak.ingress.hostname }}
-{{- else }}
-    http://{{ .Values.keycloak.ingress.hostname }}
-{{- end }}
   {{- else }}
+    http://{{ .Values.keycloak.ingress.hostname }}
```
Comment (nexeck) [inline incubator/keycloak/templates/NOTES.txt]: The hostname value does not exists anymore. It is a list with hosts.

File: incubator/keycloak/templates/NOTES.txt [hunk 2]
```diff
@@ -5,40 +5,47 @@
...
+{{- end }}
 
-  To get your user password run:
-    kubectl get secret --namespace {{ .Release.Namespace }} {{ template "keycloak.fullname" . }} -o jsonpath="{.data.keycloak-password}" | base64 --decode; echo
+2. Login with the following credentials
 
   Username: {{ .Values.keycloak.username }}
+
+  To retrieve the initial user password run:
+  kubectl get secret --namespace {{ .Release.Namespace }} {{ template "keycloak.fullname" . }} -o jsonpath="{.data.password}" | base64 --decode; echo
```
Comment (nexeck) [inline incubator/keycloak/templates/NOTES.txt]: The secret name is: { template "keycloak.fullname" . }}-http

**Review by nexeck — CHANGES_REQUESTED** [review]

**Comment by nexeck** [comment]
Keycloak with external postgresql cluster and one keycloak instance is running on k8s 1.8.1

**Comment by unguiculus** [comment]
Thanks for reviewing. Anything else?

**Comment by unguiculus** [comment]
@nexeck I improved the configuration of external databases and tested it with PostgreSQL and MariaDB. Could you test it with your setup as well?

**Comment by nexeck** [comment]
Works for me:
- External PostgreSQL
- One Replication

**Comment by nexeck** [comment]
Just had something strange with our keycloak setup.
The Statefulset started the pod and killed it after ~2 Minutes and started a new one.
This happens a few thousand times over the weekend.
But i was not able to detect the error in the keycloak logs, keycloak reported that it is up and running maybe 10-20 Seconds before the statefulset killed/restarted the pod.

**Comment by unguiculus** [comment]
Have you checked kubelet logs? Maybe you have failing liveness probes or your pods get OOMKilled or something.

**Comment by nexeck** [comment]
Will check this, if it happens again.

**Comment by nexeck** [comment]
If keycloak is running with more then one replication, do i need to set something like sticky sessions at the ingress?
Just run with more then one replication in a strange redirect loop, which says that the login token is invalid. After removing one replication everything works fine.

**Comment by nexeck** [comment]
For me, this looks fine.
We are running our keycloak instance with a single node, with the chart from this pr, since one week, without any errors, related to this chart.

**Comment by nexeck** [comment]
@unguiculus Is there anything else i could test, before this can be merged?

**Comment by unguiculus** [comment]
Did you set `keycloak.discovery` to `true` when running with more than one replica? Doing so will enable the replicated Infinispan cache. Sticky sessions should not be necessary. I tested it successfully accessing the replicas directly without a service or loadbalancer. You should see something like this in the logs:

```
keycloak-auth-1 keycloak 07:31:53,726 INFO  [org.infinispan.CLUSTER] (remote-thread--p8-t6) ISPN000310: Starting cluster-wide rebalance for cache authenticationSessions, topology CacheTopology{id=2, rebalanceId=2, currentCH=DefaultConsistentHash{ns=256, owners = (1)[keycloak-auth-1: 256+0]}, pendingCH=DefaultConsistentHash{ns=256, owners = (2)[keycloak-auth-1: 130+126, keycloak-auth-0: 126+130]}, unionCH=null, actualMembers=[keycloak-auth-1, keycloak-auth-0]}
```

**Comment by unguiculus** [comment]
I thought about that again and removed `keycloak.discovery`. High availability is now automatically enabled if `keycloak.replicas > 1`. Could you test with that again?

**Comment by clkao** [comment]
This is working nicely on 1.8.4 on gke.  Thanks for the great work.

A small issue: If the postgres persistence secret is not defined, it's auto generated.  However subsequent `helm upgrade` seems to regenerate the one used by keycloak, while leaving the one used by postgres unchanged hence keycloak fails to connect to it.

**Comment by unguiculus** [comment]
@clkao This is a common problem with all charts that auto-generated passwords. I'd suggest not to use auto-generated passwords for real-world installations. You may get around the issue using `helm upgrade --reuse-values`.

**Pull request merged**